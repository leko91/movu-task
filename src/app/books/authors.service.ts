import { Injectable } from '@angular/core';
import { of, Observable } from 'rxjs';

import { AUTHORS } from './mock-authors';
import { Author } from './author';

@Injectable({
  providedIn: 'root'
})
export class AuthorsService {

  constructor() { }

  getAuthors(): Observable<Author[]> {
    return of(AUTHORS);
  }
}
