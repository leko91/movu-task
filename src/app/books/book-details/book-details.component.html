<div class="details">
    <img width="200" height="200" src="{{ book.coverImage }}">
    <ul>
        <li>
            Author: {{ book.author }}<br>
            (Born: {{ author.dateOfBirth | date: 'dd.MM.yyyy.' }})
        </li>
        <li>
            Name: {{ book.name }}
        </li>
        <li>
            Number of pages: {{ book.numberOfPages }}
        </li>
        <li>
            Published year: {{ book.publishedYear }}
        </li>
        <li>
            <div class="input-group">
                Quantity:
                <input type="number" [(ngModel)]="bookQuantity" class="form-control">
                <div class="input-group-append">
                    <button (click)="editQuantity()" class="btn btn-secondary">Save</button>
                </div>
            </div>
        </li>
    </ul>

    <button (click)="closeDetails()" class="btn btn-secondary">Close</button>
</div>
