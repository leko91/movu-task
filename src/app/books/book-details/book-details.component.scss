$color-blue: #0061b5;

.details {
  position: absolute;
  top: 0;
  right: 0;
  z-index: 4;
  width: 300px;
  height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 30px 15px;
  color: #fff;
  background-color: darken($color-blue, 10%);

  img {
    margin: 0 0 15px;
  }
}
