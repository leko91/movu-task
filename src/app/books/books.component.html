<div class="books">
    <div class="container">
        <div class="d-flex justify-content-between align-items-center mb-3">
            <div class="dropdown">
                <button class="btn btn-secondary btn-sm dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown">Dropdown button</button>

                <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
                    <a class="dropdown-item" (click)="filterByAuthor()">Svi</a>
                    <a *ngFor="let author of authors" class="dropdown-item" (click)="filterByAuthor(author.name)">{{ author.name }}</a>
                </div>
            </div>

            <div class="input-group">
                <input type="text" [(ngModel)]="searchTerm" class="form-control">
                <div class="input-group-append">
                    <button (click)="search()" class="btn btn-outline-secondary">Search</button>
                </div>
            </div>
        </div>

        <table class="books table">
            <thead>
                <tr>
                    <th scope="col"></th>
                    <th scope="col" (click)="sortBooks('name')">Name</th>
                    <th scope="col" (click)="sortBooks('author')">Author</th>
                    <th scope="col" (click)="sortBooks('numberOfPages')">No. of pages</th>
                </tr>
            </thead>
            <tbody>
                <tr *ngFor="let book of isFiltered ? filteredBooks : books  " [routerLink]="[book.id]">
                    <td>
                        <img src="{{ book.coverImage }}">
                    </td>
                    <td>{{ book.name }}</td>
                    <td>{{ book.author }}</td>
                    <td>{{ book.numberOfPages }}</td>
                </tr>
            </tbody>
        </table>

        <ul *ngIf="totalPages > 0" class="pagination">
            <li [class.disabled]="currentPage === 1" class="page-item">
                <a class="page-link" (click)="prevPage()">Previous</a>
            </li>
            <li [class.active]="currentPage === i + 1" *ngFor="let page of ' '.repeat(totalPages).split(''), let i = index" class="page-item">
                <a (click)="goToPage(i + 1)" class="page-link">{{ i + 1 }}</a>
            </li>
            <li [class.disabled]="currentPage === totalPages" class="page-item">
                <a class="page-link" (click)="nextPage()">Next</a>
            </li>
        </ul>
    </div>
</div>

<router-outlet></router-outlet>
