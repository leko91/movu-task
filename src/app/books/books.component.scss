.books {
  padding: 30px 0;

  .input-group {
    max-width: 300px;
  }
}

.table {
  thead {
    th {
      cursor: pointer;
    }
  }

  img {
    width: 20px;
  }
}
