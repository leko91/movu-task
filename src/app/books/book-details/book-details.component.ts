import { Component, OnInit, Renderer2 } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';

import { BooksService } from '../books.service';
import { Book } from '../book';

@Component({
  selector: 'app-book-details',
  templateUrl: './book-details.component.html',
  styleUrls: ['./book-details.component.scss']
})
export class BookDetailsComponent implements OnInit {
  bookId: string;
  book: Book;
  books: Book[] = [];
  bookQuantity: number;

  constructor(
    private booksService: BooksService,
    private route: ActivatedRoute,
    private router: Router,
    private renderer: Renderer2
  ) { }

  ngOnInit() {
    this.route.paramMap.subscribe(params => {
      this.bookId = params.get('id');
    });

    this.booksService.getBooks().subscribe(res => {
      this.books = res;

      this.getBook();
    });
  }

  getBook() {
    const quantityKey = 'quantity';

    this.book = this.booksService.getBookById(this.books, this.bookId);
    this.bookQuantity = this.book[quantityKey];
    this.renderer.addClass(document.body, 'details-open');
  }

  closeDetails() {
    this.router.navigate(['../'], { relativeTo: this.route });

    this.renderer.removeClass(document.body, 'details-open');
  }

  editQuantity() {
    this.bookQuantity = this.booksService.editBookQuantity(this.books, this.bookId, this.bookQuantity);
  }
}
