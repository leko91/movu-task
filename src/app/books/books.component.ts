import { Component, OnInit } from '@angular/core';

import { BooksService } from './books.service';
import { Book } from './book';

@Component({
  selector: 'app-books',
  templateUrl: './books.component.html',
  styleUrls: ['./books.component.scss']
})
export class BooksComponent implements OnInit {
  books: Book[] = [];
  booksCopy: Book[] = [];
  filteredBooks: Book[] = [];
  filteredBooksCopy: Book[] = [];
  booksSorted = false;
  booksPerPage = 10;
  currentPage = 1;
  totalPages: number;
  isFiltered = false;
  searchTerm = '';

  constructor(
    private booksService: BooksService
  ) { }

  ngOnInit() {
    this.getBooks();
  }

  getBooks() {
    this.booksService.getBooks().subscribe(
      res => {
        this.booksCopy = res;
        this.books = this.getPaginated(res);
        this.totalPages = this.getTotalPages(this.booksCopy);
      },
      err => {
        alert(err);
      }
    );
  }

  getPaginated(inputArray) {
    return inputArray.slice((this.currentPage - 1) * this.booksPerPage, (this.currentPage) * this.booksPerPage);
  }

  getTotalPages(inputArray) {
    return Math.round(inputArray.length / this.booksPerPage);
  }

  // Basic sort
  sortBooks(key) {
    this.booksSorted = !this.booksSorted;

    let sortArray;
    if (this.isFiltered) {
      sortArray = this.filteredBooksCopy;
    } else {
      sortArray = this.booksCopy;
    }

    sortArray.sort((elem1, elem2) => {
      if (elem1[key] > elem2[key]) {
        return this.booksSorted ? 1 : -1;
      }

      if (elem1[key] < elem2[key]) {
        return this.booksSorted ? -1 : 1;
      }
    });

    if (this.isFiltered) {
      this.filteredBooks = this.getPaginated(this.filteredBooksCopy);
    } else {
      this.books = this.getPaginated(this.booksCopy);
    }
  }

  // Dropdown filter
  filterByAuthor(author) {
    if (author === undefined) {
      this.isFiltered = false;
      this.filteredBooks = [];
      this.currentPage = 1;
      this.books = this.getPaginated(this.booksCopy);
      this.totalPages = this.getTotalPages(this.booksCopy);
      return;
    }

    const keyAuthor = 'author';
    this.filteredBooks = this.booksCopy.reduce((acc, cur) => {
      if (cur[keyAuthor] === author) {
        acc.push(cur);
      }
      return acc;
    }, []);
    this.filteredBooksCopy = this.filteredBooks;
    this.isFiltered = true;
    this.currentPage = 1;
    this.totalPages = this.getTotalPages(this.filteredBooks);
    this.filteredBooks = this.getPaginated(this.filteredBooksCopy);
  }

  // Pagination
  nextPage() {
    this.currentPage++;

    if (this.isFiltered) {
      this.filteredBooks = this.getPaginated(this.filteredBooksCopy);
    } else {
      this.books = this.getPaginated(this.booksCopy);
    }
  }

  prevPage() {
    this.currentPage--;

    if (this.isFiltered) {
      this.filteredBooks = this.getPaginated(this.filteredBooksCopy);
    } else {
      this.books = this.getPaginated(this.booksCopy);
    }
  }

  goToPage(page) {
    this.currentPage = page;

    if (this.isFiltered) {
      this.filteredBooks = this.getPaginated(this.filteredBooksCopy);
    } else {
      this.books = this.getPaginated(this.booksCopy);
    }
  }

  // Search
  search() {
    const keyName = 'name';
    this.filteredBooks = this.booksCopy.reduce((acc, cur) => {
      if (cur[keyName].toLowerCase().indexOf(this.searchTerm.toLowerCase()) !== -1) {
        acc.push(cur);
      }

      return acc;
    }, []);

    if (this.filteredBooks.length === 0) {
      this.filteredBooks = [];
    }

    this.filteredBooksCopy = this.filteredBooks;
    this.isFiltered = true;
    this.currentPage = 1;
    this.totalPages = this.getTotalPages(this.filteredBooks);
    this.filteredBooks = this.getPaginated(this.filteredBooksCopy);
  }
}
