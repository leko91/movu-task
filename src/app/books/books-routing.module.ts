import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

import { BookDetailsComponent } from './book-details/book-details.component';
import { BooksComponent } from './books.component';

const booksRoutes: Routes = [
  {
    path: '',
    component: BooksComponent,
    children: [
      {
        path: ':id',
        component: BookDetailsComponent
      }
    ]
  }
];

@NgModule({
  imports: [RouterModule.forChild(booksRoutes)],
  exports: [RouterModule]
})
export class BooksRoutingModule { }
