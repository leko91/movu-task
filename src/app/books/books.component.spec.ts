import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { BooksComponent } from './books.component';
import { BookDetailsComponent } from './book-details/book-details.component';
import { RouterTestingModule } from '@angular/router/testing';
import { FormsModule } from '@angular/forms';

describe('BooksComponent', () => {
  let fixture: ComponentFixture<BooksComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        FormsModule
      ],
      declarations: [
        BooksComponent,
        BookDetailsComponent
      ],
    }).compileComponents();
  }));

  it('should display book details', () => {
    fixture = TestBed.createComponent(BooksComponent);
    expect(fixture).toMatchSnapshot();
  });
});
