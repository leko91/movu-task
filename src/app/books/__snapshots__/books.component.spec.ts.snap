// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksComponent should display book details 1`] = `
<app-books
  books={[Function Array]}
  booksCopy={[Function Array]}
  booksPerPage={[Function Number]}
  booksService={[Function BooksService]}
  booksSorted="false"
  currentPage={[Function Number]}
  filteredBooks={[Function Array]}
  filteredBooksCopy={[Function Array]}
  isFiltered="false"
  searchTerm=""
>
  <div
    class="container"
  >
    <div
      class="d-flex justify-content-between"
    >
      <div
        class="dropdown"
      >
        <button
          class="btn btn-secondary dropdown-toggle"
          data-toggle="dropdown"
          id="dropdownMenuButton"
          type="button"
        >
          Dropdown button
        </button>
        <div
          aria-labelledby="dropdownMenuButton"
          class="dropdown-menu"
        >
          <a
            class="dropdown-item"
          >
            Svi
          </a>
          <a
            class="dropdown-item"
          >
            Pera Perić
          </a>
          <a
            class="dropdown-item"
          >
            Mika Mikić
          </a>
          <a
            class="dropdown-item"
          >
            Laza Lazić
          </a>
        </div>
      </div>
      <input
        type="text"
      />
      <button>
        Search
      </button>
    </div>
    <table
      class="books table"
    >
      <thead>
        <tr>
          <th
            scope="col"
          />
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Author
          </th>
          <th
            scope="col"
          >
            No. of pages
          </th>
        </tr>
      </thead>
      <tbody>
        
      </tbody>
    </table>
    
  </div><router-outlet />
</app-books>
`;
